Add SignalWithStart

vancexu opened the pull request:
releated https://github.com/uber/cadence/pull/621

purplesusan commented on .gen/go/shared/types.go:
```diff
@@ -21524,6 +21524,498 @@ func (v *SignalExternalWorkflowExecutionInitiatedEventAttributes) GetChildWorkfl
+	return
+}
+
+// GetWorkflowIdReusePolicy returns the value of WorkflowIdReusePolicy if it is set or its
```
GetWorkflowIdReusePolicy returns the value of WorkflowIdReusePolicy if it is set or its zero value if it is unset. ->
GetWorkflowIdReusePolicy returns the value of WorkflowIdReusePolicy if it is set, or a zero value if it is not set.

purplesusan commented on .gen/go/shared/types.go:
```diff
@@ -21524,6 +21524,498 @@ func (v *SignalExternalWorkflowExecutionInitiatedEventAttributes) GetChildWorkfl
+	return
+}
+
+// GetSignalName returns the value of SignalName if it is set or its
```
GetSignalName returns the value of SignalName if it is set or its zero value if it is unset. ->
GetSignalName returns the value of SignalName if it is set, or a zero value if it is not set.

purplesusan commented on client/client.go:
```diff
@@ -104,6 +104,17 @@ type (
 		SignalWorkflow(ctx context.Context, workflowID string, runID string, signalName string, arg interface{}) error
 
+		// SignalWithStartWorkflow sends a signals to a workflow in execution, if workflow is not running or not found,
+		// it starts workflow then sends the signal in transaction.
```
suggested update:
SignalWithStartWorkflow sends a signal to a running workflow. If the workflow is not running or not found, it starts the workflow and then sends the signal in transaction.

purplesusan commented on client/client.go:
```diff
@@ -104,6 +104,17 @@ type (
 
+		// SignalWithStartWorkflow sends a signals to a workflow in execution, if workflow is not running or not found,
+		// it starts workflow then sends the signal in transaction.
+		// - workflowID, runID, signalName, signalArg are same as SignalWorkflow's parameters
```
- workflowID, runID, signalName, signalArg are same as SignalWorkflow's parameters
The workflowID, runID, signalName, and signalArg parameters are the same as those of SignalWorkflow.

purplesusan commented on client/client.go:
```diff
@@ -104,6 +104,17 @@ type (
+		// SignalWithStartWorkflow sends a signals to a workflow in execution, if workflow is not running or not found,
+		// it starts workflow then sends the signal in transaction.
+		// - workflowID, runID, signalName, signalArg are same as SignalWorkflow's parameters
+		// - options, workflow, workflowArgs are same as StartWorkflow's parameters
```
- options, workflow, workflowArgs are same as StartWorkflow's parameters ->
The workflow and workflowArgs parameters are the same as those of StartWorkflow.

mfateev commented on internal/client.go:
```diff
@@ -94,6 +94,17 @@ type (
+		// The errors it can return:
+		//  - EntityNotExistsError, if domain does not exists
+		//  - BadRequestError
+		//	- InternalServiceError
```
What's the semantic of the call when RunID is specified? For example previous workflow with this RunID has completed? Are we going to start a workflow ignoring RunID? It looks weird to me. I think we want to eliminate RunID parameter from this call as well as from the thrift API.

vancexu replied:
Removed RunID. Agree having runID here might confuse user. User now only use workflowID to find latest running workflow to signal with start.

purplesusan commented on client/client.go:
```diff
@@ -104,6 +104,17 @@ type (
+		// - workflowID, runID, signalName, signalArg are same as SignalWorkflow's parameters
+		// - options, workflow, workflowArgs are same as StartWorkflow's parameters
+		// The errors it can return:
+		//  - EntityNotExistsError, if domain does not exists
```
suggested update:
SignalWithStartWorkflow can return the following errors:
- EntityNotExistsError, if the domain does not exist

purplesusan commented on idl/github.com/uber/cadence/cadence.thrift:
```diff
@@ -298,6 +298,21 @@ service WorkflowService {
     )
 
+  /**
+  * SignalWithStartWorkflowExecution is used to ensure sending a signal event to a workflow execution.
```
SignalWithStartWorkflowExecution is used to ensure sending a signal event to a workflow execution. ->
SignalWithStartWorkflowExecution is used to send a signal event to an executed workflow.

purplesusan commented on idl/github.com/uber/cadence/cadence.thrift:
```diff
@@ -298,6 +298,21 @@ service WorkflowService {
 
+  /**
+  * SignalWithStartWorkflowExecution is used to ensure sending a signal event to a workflow execution.
+  * If workflow is running, this results in WorkflowExecutionSignaled event recorded in the history
```
this results in WorkflowExecutionSignaled event recorded in the history -> this results in a WorkflowExecutionSignaled event being recorded in the history

purplesusan commented on idl/github.com/uber/cadence/cadence.thrift:
```diff
@@ -298,6 +298,21 @@ service WorkflowService {
+  * SignalWithStartWorkflowExecution is used to ensure sending a signal event to a workflow execution.
+  * If workflow is running, this results in WorkflowExecutionSignaled event recorded in the history
+  * and a decision task being created for the execution.
+  * If workflow is not running or not found, this results in WorkflowExecutionStarted and WorkflowExecutionSignaled
```
If workflow is not running -> If the workflow is not running

purplesusan commented on idl/github.com/uber/cadence/cadence.thrift:
```diff
@@ -298,6 +298,21 @@ service WorkflowService {
+  * If workflow is running, this results in WorkflowExecutionSignaled event recorded in the history
+  * and a decision task being created for the execution.
+  * If workflow is not running or not found, this results in WorkflowExecutionStarted and WorkflowExecutionSignaled
+  * event recorded in history, and a decision task being created for the execution
```
event recorded in history, and a decision task being created for the execution -> events recorded in the history, and a decision task being created for the execution.

purplesusan commented on internal/client.go:
```diff
@@ -94,6 +94,17 @@ type (
 		SignalWorkflow(ctx context.Context, workflowID string, runID string, signalName string, arg interface{}) error
 
+		// SignalWithStartWorkflow sends a signals to a workflow in execution, if workflow is not running or not found,
+		// it starts workflow then sends the signal in transaction.
```
suggested update:
SignalWithStartWorkflow sends a signal to a running workflow. If the workflow is not running or not found, it starts the workflow and then sends the signal in transaction.

purplesusan commented on internal/client.go:
```diff
@@ -94,6 +94,17 @@ type (
 
+		// SignalWithStartWorkflow sends a signals to a workflow in execution, if workflow is not running or not found,
+		// it starts workflow then sends the signal in transaction.
+		// - workflowID, runID, signalName, signalArg are same as SignalWorkflow's parameters
```
suggested update:
The workflowID, runID, signalName, and signalArg parameters are the same as those of SignalWorkflow.

purplesusan commented on internal/client.go:
```diff
@@ -94,6 +94,17 @@ type (
+		// SignalWithStartWorkflow sends a signals to a workflow in execution, if workflow is not running or not found,
+		// it starts workflow then sends the signal in transaction.
+		// - workflowID, runID, signalName, signalArg are same as SignalWorkflow's parameters
+		// - options, workflow, workflowArgs are same as StartWorkflow's parameters
```
suggested update:
The options, workflow, and workflowArgs parameters are same as those of StartWorkflow.

purplesusan commented on internal/client.go:
```diff
@@ -94,6 +94,17 @@ type (
+		// - workflowID, runID, signalName, signalArg are same as SignalWorkflow's parameters
+		// - options, workflow, workflowArgs are same as StartWorkflow's parameters
+		// The errors it can return:
+		//  - EntityNotExistsError, if domain does not exists
```
suggested update:
SignalWithStartWorkflow can return the following errors:
- EntityNotExistsError, of the domain does not exist

purplesusan commented on internal/client.go:
```diff
@@ -94,6 +94,17 @@ type (
+		SignalWithStartWorkflow(ctx context.Context, workflowID string, runID string, signalName string, signalArg interface{},
+			options StartWorkflowOptions, workflow interface{}, workflowArgs ...interface{}) (*WorkflowExecution, error)
+
 		// CancelWorkflow cancels a workflow in execution
```
CancelWorkflow cancels a workflow in execution -> CancelWorkflow cancels a running workflow.

purplesusan commented on internal/internal_workflow_client.go:
```diff
@@ -263,6 +263,89 @@ func (wc *workflowClient) SignalWorkflow(ctx context.Context, workflowID string,
 		}, serviceOperationRetryPolicy, isServiceTransientError)
 }
 
+// SignalWithStartWorkflow sends a signals to a workflow in execution, if workflow is not running or not found,
```
suggested update:
SignalWithStartWorkflow sends a signal to a running workflow. If the workflow is not running or not found, it starts the workflow and then sends the signal in transaction.

purplesusan commented on internal/internal_workflow_client.go:
```diff
@@ -263,6 +263,89 @@ func (wc *workflowClient) SignalWorkflow(ctx context.Context, workflowID string,
+		decisionTaskTimeout = defaultDecisionTaskTimeoutInSecs
+	}
+
+	// Validate type and its arguments.
```
Validate type and its arguments. -> Validate the workflow type and its arguments.

purplesusan commented on internal/internal_workflow_client.go:
```diff
@@ -263,6 +263,89 @@ func (wc *workflowClient) SignalWorkflow(ctx context.Context, workflowID string,
+
+	var response *s.StartWorkflowExecutionResponse
+
+	// Start creating workflow request.
```
Start creating workflow request. -> Create the workflow request.

purplesusan commented on internal/internal_workflow_client_test.go:
```diff
@@ -503,3 +503,81 @@ func getGetWorkflowExecutionHistoryRequest(filterType shared.HistoryEventFilterT
 	return request
 }
+
+// workflow client test suite
```
workflow client test suite -> Workflow client test suite.

purplesusan commented on mocks/Client.go:
```diff
@@ -286,6 +286,35 @@ func (_m *Client) SignalWorkflow(ctx context.Context, workflowID string, runID s
 	return r0
 }
 
+// SignalWithStartWorkflow provides a mock function with given fields: ctx, options, workflow, args
```
SignalWithStartWorkflow provides a mock function with given fields: ctx, options, workflow, args ->
SignalWithStartWorkflow provides a mock function with the following fields: ctx, options, workflow, args.

vancexu commented on .gen/go/cadence/idl.go:
```diff
@@ -33,11 +33,11 @@ var ThriftModule = &thriftreflect.ThriftModule{
 }
 
-const rawIDL = "// Copyright (c) 2017 Uber Technologies, Inc.\n//\n// Permission is hereby granted, free of charge, to any person obtaining a copy\n// of this software and associated documentation files (the \"Software\"), to deal\n// in the Software without restriction, including without limitation the rights\n// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell\n// copie ...
+const rawIDL = "// Copyright (c) 2017 Uber Technologies, Inc.\n//\n// Permission is hereby granted, free of charge, to any person obtaining a copy\n// of this software and associated documentation files (the \"Software\"), to deal\n// in the Software without restriction, including without limitation the rights\n// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell\n// copie ...
```
dev team are advised to keep it 2017 from legal team.

vancexu commented:
Thanks @purplesusan for the suggestions. files under ./gen is auto generated by tools, so we do not have control on that. I addressed some of your comments, and keep some unchanged to keep consistency with others.
